File: Cpp/211.cpp
/*
Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)
search(word) can search a literal word or a regular expression string containing
only letters a-z or .. A . means it can represent any one letter.

Example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true
Note:
You may assume that all words are consist of lowercase letters a-z.
*/

#include <iostream>
#include <string>
#include <vector>

using namespace std;

class WordDictionary {
public:
    struct TrieNode {
        TrieNode* next[26];
        bool have;
        TrieNode(bool h = false) : have(h) {
            for (int i = 0; i < 26; ++i) {
                next[i] = nullptr;
            }
        }
    };
    TrieNode* root;
    /** Initialize your data structure here. */
    WordDictionary() { root = new TrieNode(); }

    /** Adds a word into the data structure. */
    void addWord(string word) {
        TrieNode* cur = root;
        for (int i = 0; i < word.size(); ++i) {
            if (cur->next[word[i] - 'a'] == nullptr) {
                cur->next[word[i] - 'a'] = new TrieNode();
            }
            cur = cur->next[word[i] - 'a'];
        }

        cur->have = true;
    }

    /** Returns if the word is in the data structure. A word could contain the
     * dot character '.' to represent any one letter. */
    bool search(string word) {
        TrieNode* cur = root;
        return search(word, cur, 0);
    }

    bool search(const string& word, TrieNode* cur, int pos) {
        if (cur == nullptr) {
            return false;
        }

        for (int i = pos; i < word.size(); ++i) {
            if (word[i] == '.') {
                for (int ch = 0; ch < 26; ++ch) {
                    if (search(word, cur->next[ch], i + 1)) {
                        return true;
                    }
                }
                return false;
            }

            if (cur->next[word[i] - 'a'] == nullptr) {
                return false;
            }

            cur = cur->next[word[i] - 'a'];
        }

        return cur->have;
    }

    ~WordDictionary() { delete root; }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */

int main() {}

File: Python/211.py
class TrieNode(object):
    def __init__(self):
        self.have = False
        self.next = [None]*26
        
class WordDictionary(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()
        

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: None
        """
        cur = self.root
        for c in word:
            ind = ord(c) - ord('a')
            if cur.next[ind] == None:
                cur.next[ind] = TrieNode()
            cur = cur.next[ind]
        cur.have = True
        

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        return self.searchStartFromNode(self.root, word)
    
    def searchStartFromNode(self, root, word):
        if root == None:
            return False
        cur = root
        for i in range(len(word)):
            if word[i].isalpha():
                ind = ord(word[i]) - ord('a')
                if cur.next[ind] == None:
                    return False
                cur = cur.next[ind]
            elif word[i] == ".":
                for j in range(26):
                    if self.searchStartFromNode(cur.next[j], word[i+1:]):
                        return True
                return False    
        return cur.have
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
